org.uqbar.project.wollok.lib/src/wollok.wlk

```diff
@@ -246,15 +251,35 @@ package lang {
 		 * If more than one element satisfies the condition then it depends on the specific collection class which element
 		 * will be returned
 		 * @returns the element that complies the condition
+		 * @throws ElementNotFoundException if no element matched the given predicate
 		 * Example:
 		 *      users.find { user => user.name() == "Cosme Fulanito" }
 		 */
-		method find(predicate) = this.fold(null, { acc, e =>
-			 if (acc != null)
-			 	acc
-			 else
-			 	if (predicate.apply(e)) e else null
+		method find(predicate) = this.findOrElse(predicate, { 
+			throw new ElementNotFoundException("there is no element that satisfies the predicate")
 		})
+
+		/**
+		 * Returns the element of this collection that satisfy a given condition, or the given default otherwise, if no element matched the predicate
+		 * If more than one element satisfies the condition then it depends on the specific collection class which element
+		 * will be returned
+		 * @returns the element that complies the condition or the default value
+		 * Example:
+		 *      users.findOrElse({ user => user.name() == "Cosme Fulanito" }, homer)
+		 */
+		method findOrDefault(predicate, value) =  this.findOrElse(predicate, { value })
+		
+		/**
+		 * Returns the element of this collection that satisfy a given condition, 
+		 * or the the result of evaluating the given continuation 
+		 * If more than one element satisfies the condition then it depends on the specific collection class which element
+		 * will be returned
+		 * @returns the element that complies the condition or the result of evaluating the continuation
+		 * Example:
+		 *      users.findOrElse({ user => user.name() == "Cosme Fulanito" }, { homer })
+		 */
+		method findOrElse(predicate, continuation) native
```

Comment:
I actually tried the following manners:

* using a wollok fold. That works, but it requires the whole collection to be iterated - no short circuit -, and that requires the continuation to never return null. It makes the contract more complex
* using a wollok forEach. That could work, but you can not return from the method within the closure, since closure return only escapes from the closure. It is a local return

Have I misunderstood something?